tests/test_seq2seq.py

```diff
@@ -130,6 +130,7 @@ def test_badinput(self):
 
 
 class TestHogwildSeq2seq(unittest.TestCase):
+    # os_x tests: Hangs if using pip installation of pytorch; Succeeds if using conda installation
```

Comment:
I'm so confused. If this is the case then people need to be using conda to install parlai on os x, and that's a bit annoying and undesirable.

Let's try to trace back the issue.

Reply:
Yes this has been bothering me for the past few days! It is really strange that such a thing is causing an issue. Luckily @klshuster observed that it was working locally using conda and when I tried the same it was true!
Strangely enough, this test worked on an earlier version of torch installed using pip.

How would you suggest we solve this ?

Reply:
I'm going to harass some of the pytorch devs and see what they say. Can you give specific me commands which fail/succeed?

Reply:
Sounds like the right plan!
So using a pip installation ```test_generation_multi (__main__.TestHogwildSeq2seq)``` just hangs and doesn't complete (tried locally once and no output for > 30 minutes)

Using conda the entire test file (that test and 5 more) run in about ```36.956s``` on my machine!
The command is mentioned below: 

Running from ParlAI:
```python tests/test_seq2seq.py -v```

Reply:
jumping in here to reiterate that `python tests/test_seq2seq.py` hangs with pip but completes successfully with conda (i.e. that is the command to run)

Reply:
I'd like to isolate a test case we could provide the pytorch devs to file a bug report. If @vedantpuri wants to do this, that would be awesome. We can also have him file the official report so that he gets that sweet pytorch contributor credit.

Reply:
Sure I can try doing that. Just to be clear what do you mean specifically by "isolate a test case". 
I believe that the train loop is involved in this bug so I would probably need to create a seq2seq trainer to reproduce the bug. Any thoughts how to go about it exactly ?

Reply:
We need a self-contained .py script that only imports torch, etc. The script should hang or fail when run on os x pip and succeed on os x conda. The whole thing should be very small, no more than 50 lines or so.

If we're unable to create this isolated test case, then it means we have a bug in the hogwild code in ParlAI, and we're violating some contract.

Reply:
To add on to this - we would want to find where in the seq2seq model perhaps (or the trainloop) that the code is hanging

Reply:
Do we know which pytorch version is hanging? Maybe `pip install` vs conda are causing different versions to be installed?

Reply:
Both are torch 1.2.0

Reply:
So I have been trying to get to the bottom of this for the past few hours and even though I haven't figured it out completely, I have some updates and could use some help:
- Following the trace of our test I reached ```ParlAI/parlai/core/testing_utils```.  Now within ```train_model(opt)``` everything works normal in both distributions upto: 
```python
valid, test = tl.train()
```
at which point the pip distribution hangs and conda passes through. Naturally, I dug deeper to the next layer: ```ParlAI/parlai/scripts/train_model```
***Note:*** At this point I would suggest you guys comment out ```with capture_output() as output:``` and reindent the code in order to reproduce the steps I have added below. It would also be useful to comment out all other tests other than ```TestHogwildSeq2seq```
- Now within the train function I inspected to see at which point the pip installation hangs and started adding before and after style print statements within the loop. After a while it was narrowed down to the call: ```world.parley()```. I created a counter variable outside the loop and incremented it at the bottom of the loop. Around the ```world.parley()``` statement I printed the counter, looking something like this:
```python
        ctr = 0
        with world:
            while True:
                print(ctr)
                # do one example / batch of examples
                world.parley()
                print(ctr)
                self.parleys += 1

                # get the total training examples done, compute epochs
                self._total_epochs = (
                    self._preempted_epochs
                    + num_workers() * self.world.get_total_epochs()
                )
                exs_per_epoch = self.world.num_examples()
                self._total_exs = int(np.round(self._total_epochs * exs_per_epoch))
                # and use the primary worker's timings for everything
                train_time, log_time, validate_time = sync_object(
                    (
                        self.train_time.time(),
                        self.log_time.time(),
                        self.validate_time.time(),
                    )
                )
                # check counters and timers
                if self._total_epochs >= self.max_num_epochs:
                    self.log()
                    print(
                        '[ num_epochs completed:{} time elapsed:{}s ]'.format(
                            self.max_num_epochs, train_time
                        )
                    )
                    break
                if train_time > self.max_train_time:
                    print('[ max_train_time elapsed:{}s ]'.format(train_time))
                    break
                if log_time > self.log_every_n_secs:
                    self.log()
                if (
                    validate_time > self.val_every_n_secs
                    or self._total_epochs - self.last_valid_epoch
                    >= self.val_every_n_epochs
                ):
                    stop_training = self.validate()
                    self.last_valid_epoch = self._total_epochs
                    if stop_training:
                        break
                if (
                    self.save_time.time() > self.save_every_n_secs
                    and opt.get('model_file')
                    and is_primary_worker()
                ):
                    print(
                        "[ saving model checkpoint: {}.checkpoint".format(
                            opt['model_file']
                        )
                    )
                    self.save_model('.checkpoint')
                    self.save_time.reset()
                ctr += 1
```
Surprisingly, in the **pip** installation *6* (7th iteration) is the last counter printed and it hangs after that, whereas conda runs smoothly till the end.

- Currently I have been looking at the ```parley()``` function in ```class HogwildWorld``` in ```ParlAI/parlai/core/worlds.py```.  For debugging purposes I added a few print statements to the function to make it look like:
```python
def parley(self):
        """Queue one item to be processed."""
        # schedule an example
        self.sync['queued_sem'].release()
        print("release")
        # keep main process from getting too far ahead of the threads
        # this way it can only queue up to numthreads unprocessed examples
        self.sync['threads_sem'].acquire()
        print("acquire")
        self.update_counters()
```
and observe that for every before after print of ```ctr`` , there is a ```release``` and ```acquire`` printed between them except for the ctr = 6.

A sample log of my pip installation debugging has been pasted here:
```
0
release
acquire
0
1
release
acquire
1
2
release
acquire
2
3
release
acquire
3
4
release
acquire
4
5
release
acquire
5
6
release

```
I tried debugging by adding other print statements as well but it only got more confusing due the unorderly fashion printing of multiple threads. I would want to say that it reaches a deadlock situation trying to acquire a lock that hasn't been released but the fact that the same thing works in conda stops me from saying that. I have been trying to come up with an isolated test case but have not been able to replicate a similar code flow to reproduce the bug yet. I could really use your help on this since you guys might know better than me about what's going on with the multithreading using the information I gave you. @stephenroller @klshuster